internal/controller/model_pod.go

```
@@ -0,0 +1,135 @@
+	nameMatch := int32(-1)
+	portMatch := int32(-1)
+	for _, epPort := range subset.Ports {
+		if epPort.Protocol != protocol {
+			continue
+		}
+		if name != "" && epPort.Name == name {
+			nameMatch = epPort.Port
+		}
+		if epPort.Port == targetPort {
+			portMatch = epPort.Port
+		}
+	}
+
+	if nameMatch >= 0 {
+		return nameMatch, nil
+	}
+	if portMatch >= 0 {
+		return portMatch, nil
+	}
+	return -1, errPortNotFoundInSubset
+}
+
+func (g *PodLoadBalancerModelGenerator) GenerateModel(portAssignment map[string]string) (*model.LoadBalancer, error) {
+	result := &model.LoadBalancer{}
+
+	ingressMap := map[string]model.IngressIP{}
+
+	for serviceKey, portID := range portAssignment {
+		id, _ := model.FromKey(serviceKey)
+		svc, err := g.services.Services(id.Namespace).Get(id.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		ep, err := g.endpoints.Endpoints(id.Namespace).Get(id.Name)
+		if err != nil {
+			// no endpoints exist or are not retrievable -> we ignore that for
+			// now because this may happen during bootstrapping of a service
+			continue
+		}
+		if len(ep.Subsets) < 1 {
+			// no point in doing anything with the service here
+			continue
+		}
+		// TODO: handle multiple subsets. This is tricky because our model
+		// currently does not support different ports per destination IP.
+		epSubset := ep.Subsets[0]
+
+		ingress, ok := ingressMap[portID]
+		if !ok {
+			ingressIP, err := g.l3portmanager.GetInternalAddress(portID)
+			if err != nil {
+				return nil, err
+			}
+			ingress = model.IngressIP{
+				Address: ingressIP,
+				Ports:   []model.PortForward{},
+			}
+		}
+
+		for _, svcPort := range svc.Spec.Ports {
+			targetPort := int32(svcPort.TargetPort.IntValue())
+			portName := svcPort.Name
+			if targetPort == 0 {
+				targetPort = svcPort.Port
+				portName = svcPort.TargetPort.String()
+			}
+			destinationPort, err := g.findPort(
+				&epSubset,
+				portName, targetPort, svcPort.Protocol,
+			)
+			if err != nil {
```

Is there a good reason that it is correct to ignore an error here?

---

@sebastianriese The question is, what else would we do? We could log an error instead, but making the entire operation fail because of a single misconfigured service is not desirable.

---

Okay, I am happy with logging a warning.